utilities/transactions/pessimistic_transaction.cc

```diff
@@ -559,7 +561,18 @@ Status PessimisticTransaction::TryLock(ColumnFamilyHandle* column_family,
   // any writes since this transaction's snapshot.
   // TODO(agiardullo): could optimize by supporting shared txn locks in the
   // future
-  if (skip_validate || snapshot_ == nullptr) {
+  if (!do_validate || snapshot_ == nullptr) {
+    assert(!assume_exclusive_tracked || (previously_locked && !lock_upgrade));
```

Comment:
Seems like this assert could just be in an additional else clause after "else if (lock_upgrade)" below.

Reply:
I updated the if-then clause. Let me know if the update has addressed your comment. I kept assert before the if-then clause so that we crash during the tests and the corner cases that have violated our assumptions would show up in tests.